src/common/scaled-scene-buffer.c

```diff
@@ -41,29 +41,71 @@ _cache_entry_destroy(struct scaled_scene_buffer_cache_entry *cache_entry, bool d
 	wl_list_remove(&cache_entry->link);
 	if (cache_entry->buffer) {
 		/* Allow the buffer to get dropped if there are no further consumers */
-		wlr_buffer_unlock(cache_entry->buffer);
-		if (drop_buffer) {
+		if (drop_buffer && !cache_entry->buffer->dropped) {
 			wlr_buffer_drop(cache_entry->buffer);
 		}
+		wlr_buffer_unlock(cache_entry->buffer);
 	}
 	free(cache_entry);
 }
 
+static struct scaled_scene_buffer_cache_entry *
+find_cache_for_scale(struct scaled_scene_buffer *scene_buffer, double scale)
+{
+	struct scaled_scene_buffer_cache_entry *cache_entry;
+	wl_list_for_each(cache_entry, &scene_buffer->cache, link) {
+		if (cache_entry->scale == scale) {
+			return cache_entry;
+		}
+	}
+	return NULL;
+}
+
 static void
 _update_buffer(struct scaled_scene_buffer *self, double scale)
 {
 	self->active_scale = scale;
 
 	/* Search for cached buffer of specified scale */
-	struct scaled_scene_buffer_cache_entry *cache_entry, *cache_entry_tmp;
-	wl_list_for_each_safe(cache_entry, cache_entry_tmp, &self->cache, link) {
-		if (cache_entry->scale == scale) {
-			/* LRU cache, recently used in front */
-			wl_list_remove(&cache_entry->link);
-			wl_list_insert(&self->cache, &cache_entry->link);
-			wlr_scene_buffer_set_buffer(self->scene_buffer, cache_entry->buffer);
-			return;
+	struct scaled_scene_buffer_cache_entry *cache_entry =
+		find_cache_for_scale(self, scale);
+	if (cache_entry) {
+		/* LRU cache, recently used in front */
+		wl_list_remove(&cache_entry->link);
+		wl_list_insert(&self->cache, &cache_entry->link);
+		wlr_scene_buffer_set_buffer(self->scene_buffer, cache_entry->buffer);
+		/*
+		 * If found in our local cache,
+		 * - self->width and self->height are already set
+		 * - wlr_scene_buffer_set_dest_size() has already been called
+		 */
+		return;
+	}
+
+	/* Search from other cached scaled-scene-buffers */
+	struct scaled_scene_buffer *scene_buffer;
+	wl_list_for_each(scene_buffer, &self->impl->cached_buffers, link) {
+		if (!self->impl->equal) {
+			break;
+		}
+		if (scene_buffer == self) {
+			continue;
+		}
+		if (!self->impl->equal(self, scene_buffer)) {
+			continue;
+		}
+		cache_entry = find_cache_for_scale(scene_buffer, scale);
+		if (!cache_entry) {
+			continue;
 		}
+
+		/* Ensure self->width and self->height are set correctly */
+		self->width = scene_buffer->width;
+		self->height = scene_buffer->height;
+
+		wlr_scene_buffer_set_buffer(self->scene_buffer, cache_entry->buffer);
+		wlr_scene_buffer_set_dest_size(self->scene_buffer, self->width, self->height);
+		return;
```

Comment:
I still think we should create a local cache entry when reusing a buffer. Otherwise it will not be reused once the original owner is closed. Maybe we can just do a `struct wlr_buffer *wlr_buffer = NULL` on top, set it here, remove the `return` and then do
```c
static void
_update_buffer(struct scaled_scene_buffer *self, double scale) {
	[..]
	if (!wlr_buffer) {
		/* Create new buffer, will get destroyed along the backing wlr_buffer */
		struct lab_data_buffer *buffer = self->impl->create_buffer(self, scale);
		self->width = buffer ? buffer->logical_width : 0;
		self->height = buffer ? buffer->logical_height : 0;
		wlr_buffer = &buffer->base;
	}
	if (wlr_buffer) {
		/* Ensure the buffer doesn't get deleted behind our back */
		wlr_buffer_lock(wlr_buffer);
	}
	/* Create or reuse cache entry */
	[..]
	cache_entry->buffer = wlr_buffer;
	[..]
}
```

Reply:
Yeah I agree. I didn't do it just because it will incur some overhead for iterating over caches of each buffers when creating the first buffer for a certain scale (`impl->equal()` always returns `true` but `find_cache_for_scale()` returns `NULL`). But the overhead will be negligible and I agree that we should not differentiate owner / borrower buffers.

Reply:
Done